app/admin/sponsorship.rb

```diff
@@ -1,30 +1,47 @@
 ActiveAdmin.register Sponsorship do
   menu if: proc { false }
-  actions :create, :destroy
+  actions :create
   belongs_to :sponsor
 
   member_action :inactivate, method: :put do
-    sponsorship = Sponsorship.find(params[:id])
     sponsor = Sponsor.find(params[:sponsor_id])
-    if sponsorship.inactivate params[:end_date]
-      flash[:success] = 'Sponsorship link was successfully terminated'
-    else
-      flash[:warning] = sponsorship.errors.full_messages || 'Sponsorship not terminated'
+    sponsorship = sponsor.sponsorships.find(params[:id])
+    begin
+      InactivateSponsorship.new(sponsorship: sponsorship,
+                                end_date: params[:end_date]).call
+      flash[:success] = 'Sponsorship link was successfully terminated.'
+    rescue => error
+      flash[:warning] = error.message
+    ensure
+      redirect_to admin_sponsor_path(sponsor)
+    end
+  end
+
+  member_action :destroy, method: :delete do
+    sponsor = Sponsor.find(params[:sponsor_id])
+    sponsorship = sponsor.sponsorships.find(params[:id])
+    begin
+      DestroySponsorship.new(sponsorship).call
+      flash[:success] = 'Sponsorship record was successfully destroyed.'
+    rescue => error
+      flash[:warning] = error.message
+    ensure
+      redirect_to admin_sponsor_path(sponsor)
     end
-    redirect_to admin_sponsor_path(sponsor)
   end
 
   controller do
     def create
-      @sponsor = Sponsor.find(params[:sponsor_id])
-      sponsorship= Sponsorship.new(sponsor: @sponsor, orphan_id: params[:orphan_id],
-                                      start_date: params[:sponsorship_start_date].to_s )
-      if sponsorship.save
+      sponsor = Sponsor.find(params[:sponsor_id])
+      sponsorship = sponsor.sponsorships.build(orphan_id: params[:orphan_id],
+                                               start_date: params[:sponsorship_start_date])
+      begin
+        CreateSponsorship.new(sponsorship).call
         flash[:success] = 'Sponsorship link was successfully created.'
-        redirect_to admin_sponsor_path(@sponsor.id)
-      else
-        flash[:warning]= sponsorship.errors.full_messages || 'Sponsorship not created'
-        redirect_to new_sponsorship_path(@sponsor.id, scope: 'eligible_for_sponsorship')
+        redirect_to admin_sponsor_path(sponsor)
+      rescue => error
+        flash[:warning] = error.message
+        redirect_to new_sponsorship_path(sponsor, scope: 'eligible_for_sponsorship')
       end
```

Comment:
The raise-rescue approach here was taken because of AR transactions that wrap the code in `#call`. For a transaction to be rolled back in case something goes wrong, the methods within it have to throw errors rather than return `false`. It is possible to rescue within `#call` itself and return `false` to the controller, but then propagating the error messages to the flash would be more convoluted.

Constant reliance on `rescue` is not my preferred option either, perhaps we can find something better.